addons/googledrive/models.py

```diff
@@ -0,0 +1,243 @@
+# -*- coding: utf-8 -*-
+"""Persistence layer for the google drive addon.
+"""
+import os
+import urllib
+
+from django.db import models
+
+from framework.auth import Auth
+from framework.exceptions import HTTPError
+
+from addons.base.models import (
+    BaseOAuthNodeSettings, BaseOAuthUserSettings, BaseStorageAddon)
+from osf.models.external import ExternalProvider
+from osf.models.files import FileNode, Folder, File
+from website.addons.base import exceptions
+from website.addons.googledrive import settings as drive_settings
+from website.addons.googledrive.serializer import GoogleDriveSerializer
+from website.addons.googledrive.client import GoogleAuthClient, GoogleDriveClient
+from website.addons.googledrive.utils import to_hgrid
+from website.util import api_v2_url
+# from website.files.models.ext import PathFollowingFileNode
+
+
+# TODO make googledrive "pathfollowing"
+# A migration will need to be run that concats
+# folder_path and filenode.path
+# class GoogleDriveFileNode(PathFollowingFileNode):
+class GoogleDriveFileNode(FileNode):
+    provider = 'googledrive'
+    FOLDER_ATTR_NAME = 'folder_path'
+
+class GoogleDriveFolder(GoogleDriveFileNode, Folder):
+    pass
+
+class GoogleDriveFile(GoogleDriveFileNode, File):
+    pass
+
+
+class GoogleDriveProvider(ExternalProvider):
+    name = 'Google Drive'
+    short_name = 'googledrive'
+
+    client_id = drive_settings.CLIENT_ID
+    client_secret = drive_settings.CLIENT_SECRET
+
+    auth_url_base = '{}{}'.format(drive_settings.OAUTH_BASE_URL, 'auth?access_type=offline&approval_prompt=force')
+    callback_url = '{}{}'.format(drive_settings.API_BASE_URL, 'oauth2/v3/token')
+    auto_refresh_url = callback_url
+    refresh_time = drive_settings.REFRESH_TIME
+    expiry_time = drive_settings.EXPIRY_TIME
+
+    default_scopes = drive_settings.OAUTH_SCOPE
+    _auth_client = GoogleAuthClient()
+    _drive_client = GoogleDriveClient()
+
+    def handle_callback(self, response):
+        client = self._auth_client
+        info = client.userinfo(response['access_token'])
+        return {
+            'provider_id': info['sub'],
+            'display_name': info['name'],
+            'profile_url': info.get('profile', None)
+        }
+
+    def fetch_access_token(self, force_refresh=False):
+        self.refresh_oauth_key(force=force_refresh)
+        return self.account.oauth_key
+
+
+class UserSettings(BaseStorageAddon, BaseOAuthUserSettings):
+    oauth_provider = GoogleDriveProvider
+    serializer = GoogleDriveSerializer
+
+
+class NodeSettings(BaseStorageAddon, BaseOAuthNodeSettings):
+    oauth_provider = GoogleDriveProvider
+    provider_name = 'googledrive'
+
+    folder_id = models.TextField(default=None, null=True, blank=True)
```

Comment:
Django used to give a warning about that, if it doesn't anymore then being explicit might be better.